app/src/main/java/app/musikus/ui/MainActivity.kt

```diff
@@ -139,4 +94,133 @@ class MainActivity : PermissionCheckerActivity() {
 
     }
 
+    private fun initializeExportImportLaunchers() {
+
+        exportLauncher = registerForActivityResult(
+            ExportDatabaseContract()
+        ) { uri ->
+            if (uri == null) {
+                Toast.makeText(this, "No file selected", Toast.LENGTH_SHORT).show()
+                return@registerForActivityResult
+            }
+
+            contentResolver.openOutputStream(uri)?.let { outputStream ->
+                database.export(outputStream)
+                outputStream.close()
+            }
+
+            Toast.makeText(this, "Backup successful", Toast.LENGTH_LONG).show()
+
+            // Restart the app to allow dagger hilt to reopen the database
+            triggerRestart()
+        }
+
+        importLauncher = registerForActivityResult(
+            ImportDatabaseContract()
+        ) { uri ->
+            if (uri == null) {
+                Toast.makeText(this, "No file selected", Toast.LENGTH_SHORT).show()
+                return@registerForActivityResult
+            }
+
+            // Close the database to collect all logs
+            database.close()
+
+            // Create a backup of the current database before loading the new one
+            val backupFile = File(filesDir, DATABASE_IMPORT_BACKUP_FILE)
+            database.databaseFile.copyTo(backupFile, overwrite = true)
+
+            var message = "Error loading backup. Aborting..."
+
+            lifecycleScope.launch {
+                try {
+                    // Load the new database
+                    contentResolver.openInputStream(uri)?.let { inputStream ->
+                        database.import(inputStream)
+                        inputStream.close()
+                    }
+
+                    // Open the new database locally to check if the import was successful
+                    val newDatabase = MusikusDatabase.buildDatabase(context = this@MainActivity)
+
+                    val isNewDatabaseValid = newDatabase.validate()
+
+                    newDatabase.close()
+
+                    if (isNewDatabaseValid) {
+                        message = "Backup successfully loaded"
+                    } else {
+                        // If the import was invalid, throw an error to restore the backup
+                        message = "Invalid database. Restoring backup..."
```

Comment:
Meybe it's a bit confusing that you are referring to both databses as the "backup". But here, it is actually the "old" database and above it's the actual "backup" the user did in the past.

Maybe rephrase this message with only "Invalid database" (supressing the restoring because it's transparent for the user anyways) or "Invalid database. No changes made."